feat: Add user roles, but do not yet enforce them

Emyrk opened the pull request:
# What this does

Adds some builtin roles and assigns them to users and org members. This does not implement enforcing roles at this time.

Granting user's roles was added to test the api. See how to grant roles from the SDK. The "strange" part is that the `rbac` library is used to find the role name. In practice, there will be an endpoint to list all roles that can be granted, and the user will select from a list. The code for this exists, but I want to add this in future PRs as I work with the FE on endpoint format.

https://github.com/coder/coder/blob/a6436708df68b1aa334b18c0be0686ef678e1b2c/coderd/users_test.go#L326-L330

# Future Work

- Add ability to remove roles
- Add proper authorize usage on endpoints
- List roles that can be granted

Emyrk committed:
WIP

Emyrk committed:
chore: Rework roles to be expandable by name alone

Emyrk committed:
Add role migrations

Emyrk committed:
Make gen

Emyrk committed:
Add admin role to the first user

Emyrk committed:
Implement builtin roles

Emyrk committed:
Add fake for GrantRoles

Emyrk committed:
Correct indentation on sql

Emyrk committed:
Add endpoints to grant new roles to a user

Currently can't remove roles. So be careful

Emyrk committed:
Make gen after rebase

Emyrk committed:
re insert code lost in merge

Emyrk committed:
Appease the linter

Emyrk committed:
Make gen

Emyrk committed:
Audit log track rbac_roles

kylecarbs reviewed (commented):
All minor things. I'll approve once we resolve some of the outstanding q's.

Emyrk committed:
PR comments

- Remove string type alias, use string
- Drop unused funcs
- Fix down migration

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/user_assign_role_rebase

Emyrk committed:
Add orgids to user response

Emyrk committed:
Rename type to use proper casing

Emyrk committed:
feat: Split org/site roles

- UpdateRoles vs Grant/RemoveRoles
- UpdateRoles and UpdateMemberRoles

Emyrk committed:
Add more unit testing

Emyrk committed:
Update comment

kylecarbs reviewed (approved):
The separation of user and organization roles in the API feels really nice. Great work 🥳🥳🥳

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/user_assign_role_rebase
